feat(progress-bar): add Progress Bar

paulstelzer opened the pull request:
#### Short description of what this resolves:
Adds a progress bar component to Ionic 4. Animation is based on MD spec (https://material.io/design/components/progress-indicators.html#linear-progress-indicators )

**Feature Request**: #16558

BASIC
---

![2018-12-09_10-25-36](https://user-images.githubusercontent.com/20501666/49695539-fcd3a980-fb9c-11e8-8be7-d76a4ca2015d.gif)

Preview
---

![2018-12-09_10-26-16](https://user-images.githubusercontent.com/20501666/49695540-fcd3a980-fb9c-11e8-92d0-afe4b16fe512.gif)

manucorporat commented:
I like it a lot! we could definitely merge it after 4.0 final is released!

We try to avoid nested SCSS rules, could you try to flatten that to at most 1 level nested?

paulstelzer commented:
@manucorporat So I've added some things from this PR #12010, but have some questions:

1) He is using buffer-circles as svg. I've added it, but I don't like it (the colors of the buffer circles are not based on css vars at the moment - i only would integrate it if we keep it). Does anyone has a better solution for something like this?

2) Inside the render() method I added

```
        <div class="buffer-circles"></div>
        <div class="buffer-bar" style={{ width: `${this.buffer}%` }}></div>
        <div class="buffer" style={{ width: `${this.buffer}%` }}></div>
```
Is it possible to include this only if the indicator is "buffer"? I didn't get it work :(

3) Anyone should say if the properties are okay (indicator, value, buffer). Mode and color are from Ionic

4) I added 5 css variables, someone should check if the naming is okay

```
     * @prop --bar-background: Background of the progress bar
     * @prop --buffer-background: Background of the buffer bar
     * @prop --height: Height of the progress bar
     * @prop --border-radius: Border Radius of the progress bar
     * @prop --bar-speed: Speed of the progress bar
```

5) Is the scss okay so? I still use nested SCSS rules, but think it's better now

6) I don't have different styles for ios and md mode. Is it okay?

7) I did not added the e2e.ts because I don't know if I can just copy it from another component, so anyone else can please add it

And so it looks right now:

![2018-12-03_13-11-46](https://user-images.githubusercontent.com/20501666/49375336-4be48f00-f704-11e8-892d-c68e72853fce.gif)

brandyscarney commented:
Very nice! I'll review this soon.

paulstelzer commented on core/src/components/progress-bar/progress-bar.scss:
```diff
@@ -0,0 +1,108 @@
+    --bar-background: rgba(0, 0, 0, .12);
+    overflow: hidden;
+    box-sizing: border-box;
+    display: block;
```
done

paulstelzer commented:
I've updated the progress-bar, but one issue I cannot fix:

```
src/components/progress-bar/progress-bar.scss
 145:9  ×  Unexpected property "left"   property-blacklist
 150:9  ×  Unexpected property "left"   property-blacklist
 158:9  ×  Unexpected property "left"   property-blacklist
 163:9  ×  Unexpected property "left"   property-blacklist
```

left is needed for the animation and I tried to simplify it, but I still need left :(

paulstelzer commented:
So, I've rewirtten a lot after discussion with @manucorporat - The animation is now using MD spec -> https://material-components.github.io/material-components-web-catalog/#/component/linear-progress-indicator

manucorporat commented on core/src/components/progress-bar/progress-bar.scss:
```diff
@@ -0,0 +1,232 @@
+
+// For indeterminate and query seperate bars will be needed
+// to add the MD based animation
+:host(.indeterminate),
```
not needed

manucorporat replied:
.primary-bar and friends are only used when the type is query or indeterminate anyway, it does not need to be wrapped around `:host(.indeterminate),`

manucorporat commented on core/src/components/progress-bar/progress-bar.scss:
```diff
@@ -0,0 +1,232 @@
+}
+
+// On reversed the bar starts on the right
+:host(.reversed) {
```
make flat rule

manucorporat commented on core/src/components/progress-bar/progress-bar.scss:
```diff
@@ -0,0 +1,232 @@
+:host(.reversed),
+:host(.determinate) {
+    .progress {
+        background: currentColor;
```
not needed move, into .progress

manucorporat commented on core/src/components/progress-bar/progress-bar.scss:
```diff
@@ -0,0 +1,232 @@
+    }
+}
+
+:host(.query) {
```
flatten

manucorporat commented on core/src/components/progress-bar/progress-bar.scss:
```diff
@@ -0,0 +1,232 @@
+    }
+}
+
+:host(.buffer) {
```
flat

manucorporat commented on core/src/components/progress-bar/progress-bar.scss:
```diff
@@ -0,0 +1,232 @@
+
+:host(.buffer) {
+    background: none;
+    .buffer-background {
```
this rules does not need to be around :host(.buffer)

manucorporat commented on core/src/components/progress-bar/progress-bar.tsx:
```diff
@@ -0,0 +1,78 @@
+})
+export class ProgressBar implements ComponentInterface {
+
+  @Element() el!: HTMLElement;
```
not needed

manucorporat commented on core/src/components/progress-bar/progress-bar.tsx:
```diff
@@ -0,0 +1,78 @@
+   * Sets the indicator style of the progress bar
+   * Options are determinate (no animation), reversed (reversed determinate), indeterminate (animate from left to right) and query (animate from right to left)
+   */
+  @Prop() indicator: 'reversed' | 'determinate' | 'indeterminate' | 'query' | 'buffer' = 'determinate';
```
should not reversed by another prop? reversed could be applied to any mode.

manucorporat replied:
rename to "type"

manucorporat commented on core/src/components/progress-bar/progress-bar.tsx:
```diff
@@ -0,0 +1,78 @@
+    return {
+      class: {
+        ...createColorClasses(this.color),
+        [this.indicator]: true,
```
prefix with `progress-bar-`

manucorporat commented on core/src/components/progress-bar/progress-bar.tsx:
```diff
@@ -0,0 +1,78 @@
+    const content = [];
+    if (this.indicator === 'indeterminate' || this.indicator === 'query') {
+      content.push(
+        <div class="progress">
```
`<div class="progress">` is not needed, remove from here

manucorporat commented on core/src/components/progress-bar/progress-bar.tsx:
```diff
@@ -0,0 +1,84 @@
+        <div class="indeterminate-bar-secondary"><span class="progress-indeterminate"></span></div>
+      );
+    } else {
+      const percent = this.value / 100;
```
- 0 to 1, feels more "standard" to me, arguments against it?
- we should clamp the value to the valid range [0, 1] or [0, 100]
https://github.com/ionic-team/ionic/blob/master/core/src/utils/helpers.ts#L47

paulstelzer replied:
I also thought about it, but I think 0...100 is better (angular material is also using this way) and it feels more natural https://material.angular.io/components/progress-bar/overview

manucorporat replied:
Looks like Apple uses [0, 1] in UIProgressView: https://developer.apple.com/documentation/uikit/uiprogressview/1619846-setprogress
Material Web Components uses [0, 1]: https://material.io/develop/web/components/linear-progress/
Angular uses [0, 100] https://developer.android.com/reference/android/widget/ProgressBar

paulstelzer replied:
Clamp is now integrated for value and buffer

manucorporat commented on core/src/components/progress-bar/progress-bar.scss:
```diff
@@ -0,0 +1,208 @@
+    --buffer-background: rgba(0, 0, 0, .30);
+    --buffer-color: currentColor;
+    --buffer-size: 10px;
+    --buffer-speed: 450ms;
```
Not really a speed, but a duration. I think it's better to be more explicit here, how about:
`--buffer-animation-duration`?

manucorporat commented on core/src/components/progress-bar/test/basic/index.html:
```diff
@@ -0,0 +1,143 @@
+        }
+
+        ion-progress-bar {
+            margin: 10px 0;
```
I think it's important for ionic component to look nice by default, should we include this CSS in the host element, directly? it could still be override.

paulstelzer replied:
The question is where to set the progress bar - if I want it at the top, then it would be great to have no margin by default - and in my use cases I dont have a margin for a progress bar

brandyscarney commented on core/src/components/progress-bar/progress-bar.tsx:
```diff
@@ -0,0 +1,82 @@
+  @Prop() value = 0;
+
+  /**
+   * shows circle points if buffer and value are smaller than 1 - Value of the buffer from [0, ..., 1]
```
Can we capitalize this :grin:

brandyscarney commented on core/src/components/progress-bar/readme.md:
```diff
@@ -0,0 +1,70 @@
+# ion-progress-bar
+
+ion-progress-bar is a horizontal progress bar to visualize the progression of an operation and activity. You can choose between two types: ``determinate`` and ``indeterminate``.
```
Remove the double code ticks here please

brandyscarney commented on core/src/components/progress-bar/readme.md:
```diff
@@ -0,0 +1,70 @@
+
+## Progress modes
+
+### determinate (default)
```
Can we capitalize the headers here

```
### Determinate
### Indeterminate
```

And then remove the `(default)` in the header in favor of saying it in the paragraph below. It will just make it easier to link to this in the docs like: `ionicframework.com/docs/api/progress-bar#determinate`

brandyscarney commented on core/src/components/progress-bar/readme.md:
```diff
@@ -0,0 +1,70 @@
+
+ion-progress-bar is a horizontal progress bar to visualize the progression of an operation and activity. You can choose between two types: ``determinate`` and ``indeterminate``.
+
+## Progress modes
```
Can we rename this to match the property name: `Progress Type`

brandyscarney commented:
This is looking really good! I added some comments for places the docs could improve but it isn't critical. 🙂 

My only concern with the API is using `type`, do you think we will have issues with this interfering with native? @manucorporat @adamdbradley. I believe we were calling it `indicator` in the v3 one.

manucorporat commented:
@brandyscarney i recommended to use type, because it's the obvious API, easier to remember imo.
We also use `type` in many components, ripple-effect, menu

paulstelzer commented:
@brandyscarney Thanks for your review! What about the css variables? @manucorporat and I discussed a lot of it yesterday, at the end, this is my favorite (it's not in this PR already):

```
* @prop --progress-bar-color: Color of the progress bar (foreground)
* @prop --buffer-bar-color: Color of the buffer bar (background)
* @prop --buffer-circle-color: Color of the buffer circles
```

manucorporat commented:
@paulstelzer @brandyscarney I would:
1. remove --buffer-circle-color: start with a small API, does not seem like a common use case, we can add it later
2. Add --color, and --background:

```css
  --color: #{ion-color(primary, base)};
  --background: #{ion-color(primary, base, 0.2)};
  --progress-bar-background: var(--color);
  --buffer-bar-background: var(--background);
```

brandyscarney commented:
I think we should keep it simple like the following:

```
--background: #{ion-color(primary, base, 0.2)};

--progress-background:  #{ion-color(primary, base)};
--buffer-background: var(--background);
```

which would allow for things like this to come in later:

```
--progress-background-disabled
--buffer-background-disabled
```

Also, avoid using `bar` since material design also has a circular indicator and I think these could be shared with it.

manucorporat commented on core/src/components/progress-bar/progress-bar.scss:
```diff
@@ -0,0 +1,227 @@
+  // Otherwise the buffer circles would be seen through
+  background: #fff;
+
+  z-index: -1; // Make it behind the progress
```
```suggestion
  z-index: 1; // Make it behind the progress
```

manucorporat commented on core/src/components/progress-bar/progress-bar.scss:
```diff
@@ -0,0 +1,227 @@
+  background: radial-gradient(ellipse at center, var(--buffer-background) 0%, var(--buffer-background) 30%, transparent 30%) repeat-x 5px center;
+  background-size: 10px 10px;
+
+  z-index: -1;
```
```suggestion
  z-index: 0;
```

manucorporat commented on core/src/components/progress-bar/progress-bar.scss:
```diff
@@ -0,0 +1,227 @@
+// --------------------------------------------------
+.progress,
+.progress-indeterminate {
+  background: var(--progress-background);
```
```suggestion
  background: var(--progress-background);
  
  z-index: 2;
```
